### PR TITLE
[DEPENDS ON #1594] Add `cache` to GitHub Actions Workflows

### PR DESCRIPTION
**Pull request by crcrpar**
## Motivation
Though with the cache we have not seen any great performance gain, I send this hoping the cache will reduce the pip installation failures that we've seen last couple weeks.

### CODE REVIEW

**Comment by crcrpar** [comment]
As `cache` would skip the download of pip installation which failed the last couple weeks.

File: .github/workflows/examples.yml [hunk 1]
```diff
@@ -19,10 +19,21 @@ jobs:
...
-    - name: Install
+    - name: Setup cache
+      uses: actions/cache@v2
+      env:
+        cache-name: daily-example
+      with:
+        path: ~/.cache/pip
+        key: ${{ runner.os }}-${{ matrix.python-version }}-${{ env.cache-name }}-${{ hashFiles('**/setup.py') }}-v1
+        restore-keys: |
+          ${{ runner.os }}-${{ matrix.python-version }}-${{ env.cache-name }}-${{ env.cache-name }}
```
Comment (hvy) [inline .github/workflows/examples.yml:30]: Suppose you meant the following.
```suggestion
          ${{ runner.os }}-${{ matrix.python-version }}-${{ env.cache-name }}-${{ hashFiles('**/setup.py') }}
```

File: .github/workflows/examples.yml [hunk 2]
```diff
@@ -19,10 +19,21 @@ jobs:
       uses: actions/setup-python@v2
       with:
         python-version: ${{ matrix.python-version }}
-    - name: Install
+    - name: Setup cache
+      uses: actions/cache@v2
+      env:
+        cache-name: daily-example
```
Comment (hvy) [inline .github/workflows/examples.yml:25]: What happens if we don't distinguish by workflow, does it still work?

Assuming we have to, a bit curious whether you could use some of the [default envs](https://docs.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables#default-environment-variables), such as `GITHUB_ACTION`. Not a request but wanted to know if you've tried.

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -37,6 +37,15 @@ jobs:
       with:
         python-version: ${{ matrix.python-version }}
 
+    - name: Setup Cache
```
Comment (hvy) [inline .github/workflows/tests.yml:40]: One huge nit, for consistency with your other changes.
```suggestion
    - name: Setup cache
```

File: .github/workflows/examples.yml (hunk 2 above)
Comment (hvy) [reply .github/workflows/examples.yml:25]: Ah, although `setup.py` is identical, different libraries are installed for different workflows.

**Pull request merged**